src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -202,14 +468,13 @@ bool CryptoObjectDispatch<I>::compare_and_write(
                  << dendl;
   ceph_assert(m_crypto != nullptr);
 
-  uint64_t image_offset = Striper::get_file_offset(
-          m_image_ctx->cct, &m_image_ctx->layout, object_no, object_off);
-  auto r = m_crypto->encrypt(&cmp_data, image_offset);
-  if (r == 0) {
-    r = m_crypto->encrypt(&write_data, image_offset);
-  }
-  *dispatch_result = io::DISPATCH_RESULT_CONTINUE;
-  on_dispatched->complete(r);
+  *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
+  auto req = new C_UnalignedObjectWriteRequest<I>(
+          m_image_ctx, m_crypto, object_no, object_off, &write_data, &cmp_data,
```

Comment:
Nit: I would just `std::move` the `write_data` and `cmp_data` and have the helper struct keep the original versions unmodified in case of restart.